tooling/templatize/cmd/run/options.go

```diff
@@ -0,0 +1,112 @@
+package run
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+
+	options "github.com/Azure/ARO-HCP/tooling/templatize/cmd"
+	"github.com/Azure/ARO-HCP/tooling/templatize/pkg/config"
+	"github.com/Azure/ARO-HCP/tooling/templatize/pkg/pipeline"
+)
+
+func DefaultOptions() *RawRunOptions {
+	return &RawRunOptions{
+		RolloutOptions: options.DefaultRolloutOptions(),
+	}
+}
+
+func BindOptions(opts *RawRunOptions, cmd *cobra.Command) error {
+	err := options.BindRolloutOptions(opts.RolloutOptions, cmd)
+	if err != nil {
+		return fmt.Errorf("failed to bind options: %w", err)
+	}
+	cmd.Flags().StringVarP(&opts.PipelineFile, "pipeline-config", "p", opts.PipelineFile, "pipeline file path")
+
+	for _, flag := range []string{"pipeline-config"} {
+		if err := cmd.MarkFlagFilename(flag); err != nil {
+			return fmt.Errorf("failed to mark flag %q as a file: %w", flag, err)
+		}
+	}
+	return nil
+}
+
+// RawRunOptions holds input values.
+type RawRunOptions struct {
+	RolloutOptions *options.RawRolloutOptions
+	PipelineFile   string
+}
+
+// validatedRunOptions is a private wrapper that enforces a call of Validate() before Complete() can be invoked.
+type validatedRunOptions struct {
+	*RawRunOptions
+	*options.ValidatedRolloutOptions
+}
+
+type ValidatedRunOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*validatedRunOptions
+}
+
+// completedRunOptions is a private wrapper that enforces a call of Complete() before config generation can be invoked.
+type completedRunOptions struct {
+	*ValidatedRunOptions
+	RolloutOptions *options.RolloutOptions
+	Pipeline       *pipeline.Pipeline
+}
+
+type RunOptions struct {
+	// Embed a private pointer that cannot be instantiated outside of this package.
+	*completedRunOptions
+}
+
+func (o *RawRunOptions) Validate() (*ValidatedRunOptions, error) {
+	validatedRolloutOptions, err := o.RolloutOptions.Validate()
+	if err != nil {
+		return nil, err
+	}
+	// todo validate here
+	return &ValidatedRunOptions{
+		validatedRunOptions: &validatedRunOptions{
+			RawRunOptions:           o,
+			ValidatedRolloutOptions: validatedRolloutOptions,
+		},
+	}, nil
+}
+
+func (o *ValidatedRunOptions) Complete() (*RunOptions, error) {
+	completed, err := o.ValidatedRolloutOptions.Complete()
+	if err != nil {
+		return nil, err
+	}
+
+	pipeline, err := pipeline.LoadPipeline(o.PipelineFile, completed.Config)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load pipeline file %s: %w", o.PipelineFile, err)
+	}
+
+	return &RunOptions{
+		completedRunOptions: &completedRunOptions{
+			ValidatedRunOptions: o,
+			RolloutOptions:      completed,
+			Pipeline:            pipeline,
+		},
+	}, nil
+}
+
+func (o *RunOptions) RunPipeline() error {
+	variables, err := o.RolloutOptions.Options.ConfigProvider.GetVariables(
+		o.Cloud,
+		o.DeployEnv,
+		o.Region,
```

Comment:
I haven't read the code for this yet but at first blush it's surprising that you need to pass `o.Region` to `GetVariables()` as well as to `NewEvaluationContext()` - is it valid to pass two different regions?

Reply:
```
GetVariables(cloud, deployEnv, region string, configEvalContext *ConfigEvaluationContext) (Variables, error)
```
* the region that is directly passed into `GetVariables` determines what regional overrides we want to be considered when returning config variables
* the region that is part of the configEvalContext is the one use as replacement value for `{{ .ctx.region }}` in config preprocessing

while in DEV both are the same, for EV2 they are different:
* configEvalContext.region is `$location()`
* the other is an actual region name

 i renamed `ConfigEvaluationContext` to `ConfigReplacements` and also renamed its fields to make it more apparent what they are about.